source/mesh_refinement/slope.cc

```diff
@@ -0,0 +1,113 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+
+#include <aspect/mesh_refinement/slope.h>
+
+namespace aspect
+{
+  namespace MeshRefinement
+  {
+    template <int dim>
+    void
+    Slope<dim>::execute(Vector<float> &indicators) const
+    {
+      indicators = 0;
+
+      QMidpoint<dim-1> quadrature;
+      UpdateFlags update_flags = UpdateFlags(update_normal_vectors | update_quadrature_points );
+      FEFaceValues<dim> fe_face_values (this->get_mapping(), this->get_fe(), quadrature, update_flags);
+
+      // iterate over all of the cells, get a face normal, then
+      // dot it with the local gravity
+
+      typename DoFHandler<dim>::active_cell_iterator
+      cell = this->get_dof_handler().begin_active(),
+      endc = this->get_dof_handler().end();
+
+      unsigned int i=0;
+      for (; cell!=endc; ++cell, ++i)
+        if (cell->is_locally_owned() && cell->at_boundary())
+          for (unsigned int face_no=0; face_no<GeometryInfo<dim>::faces_per_cell; ++face_no)
+            if (cell->face(face_no)->at_boundary())
+              {
+                const types::boundary_id boundary_indicator
+#if DEAL_II_VERSION_GTE(8,3,0)
+                  = cell->face(face_no)->boundary_id();
+#else
+                  = cell->face(face_no)->boundary_indicator();
+#endif
+                if ( free_surface_indicators.find(boundary_indicator) !=
+                     free_surface_indicators.end() )
+                  {
+                    fe_face_values.reinit(cell, face_no);
+ 
+                    const Tensor<1,dim> normal( fe_face_values.normal_vector(0) ); //Only one q point
+                    const Point<dim> midpoint = fe_face_values.quadrature_point(0);
+                    const Tensor<1,dim> gravity = this->get_gravity_model().gravity_vector(midpoint);
+
+                    indicators(i) = std::min ( std::acos ( normal * gravity / gravity.norm() ),
+                                               std::acos (-normal * gravity / gravity.norm() ) ); //Don't care whether gravity is in the opposite direction
+                    break;  //no need to loop over the rest of the faces
+                  }
+              }
+
+    }
+
+    template <int dim>
+    void
+    Slope<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Model settings");
+      {
+        const std::vector<types::boundary_id> x_free_surface_boundary_indicators
+          = this->get_geometry_model().translate_symbolic_boundary_names_to_ids(Utilities::split_string_list
+                                                                    (prm.get ("Free surface boundary indicators")));
```

Comment:
I'm not too fond of this -- it duplicated information in different places. I'd much rather you not have to store this piece of information in your model here, and instead introduce a function into `SimulatorAccess` that returns the set of free surface boundary indicators. This way you could also get rid of the member variable of the current class.